velox/functions/lib/DateTimeFormatter.cpp

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/lib/DateTimeFormatter.h"
+#include <date/date.h>
+#include <date/tz.h>
+#include "velox/functions/lib/DateTimeFormatterBuilder.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+
+constexpr std::string_view weekdaysFull[] = {
+    "Sunday",
+    "Monday",
+    "Tuesday",
+    "Wednesday",
+    "Thursday",
+    "Friday",
+    "Saturday"};
+constexpr std::string_view weekdaysShort[] =
+    {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
+constexpr std::string_view monthsFull[] = {
+    "January",
+    "February",
+    "March",
+    "April",
+    "May",
+    "June",
+    "July",
+    "August",
+    "September",
+    "October",
+    "November",
+    "December",
+};
+constexpr std::string_view monthsShort[] = {
+    "Jan",
+    "Feb",
+    "Mar",
+    "Apr",
+    "May",
+    "Jun",
+    "Jul",
+    "Aug",
+    "Sep",
+    "Oct",
+    "Nov",
+    "Dec",
+};
+
+template <typename T>
+std::string padContent(
+    const T& content,
+    char padding,
+    size_t totalDigits,
+    bool padFront = true) {
+  std::string strContent = folly::to<std::string>(content);
+  auto contentLength = strContent.size();
+  if (contentLength == 0) {
+    return std::string(totalDigits, padding);
+  }
+
+  bool isNegative = strContent[0] == '-';
+  auto digitLength = contentLength - (isNegative ? 1 : 0);
+  if (digitLength >= totalDigits) {
+    return strContent;
+  }
+  std::string paddingStr(totalDigits - digitLength, padding);
+  if (padFront) {
+    return strContent.insert((isNegative ? 1 : 0), paddingStr);
+  } else {
+    return strContent.append(paddingStr);
+  }
+}
+
+template <>
+std::string padContent<std::string>(
+    const std::string& content,
```

Comment:
I see somewhere that Velox actually discourages use of std::string and prefer std::string_view. Here and elsewhere

Reply:
Sure I can change them to string_view.

Reply:
I'll change the public interfaces to string_view but probably not this one. This one will only be used inside the file and creating a string_view for the purpose of using string_view as param type is perhaps overkill here in this case.